component/pyroscope/java/loop.go

```diff
@@ -0,0 +1,254 @@
+	profiles, err := jfrpprof.ParseJFR(jfrBytes, &jfrpprof.ParseInput{
+		StartTime:  startTime,
+		EndTime:    endTime,
+		SampleRate: int64(p.cfg.SampleRate),
+	}, new(jfrpprof.LabelsSnapshot))
+	if err != nil {
+		return fmt.Errorf("failed to parse jfr: %w", err)
+	}
+	for _, req := range profiles.Profiles {
+		metric := req.Metric
+		sz := req.Profile.SizeVT()
+		l := log.With(p.logger, "metric", metric, "sz", sz)
+		ls := labels.NewBuilder(nil)
+		for _, l := range jfrpprofPyroscope.Labels(p.target, profiles.JFREvent, req.Metric, "", spyName) {
+			ls.Set(l.Name, l.Value)
+		}
+		profile, err := req.Profile.MarshalVT()
+		if err != nil {
+			_ = l.Log("msg", "failed to marshal profile", "err", err)
+			continue
+		}
+		samples := []*pyroscope.RawSample{{RawProfile: profile}}
+		err = p.output.Appender().Append(context.Background(), ls.Labels(), samples)
+		if err != nil {
+			_ = l.Log("msg", "failed to push jfr", "err", err)
+			continue
+		}
+		_ = l.Log("msg", "pushed jfr-pprof")
+	}
+	return nil
+}
+
+func (p *profilingLoop) start() error {
+	p.startTime = time.Now()
+	argv := make([]string, 0, 14)
+	argv = append(argv,
+		"-f", p.jfrFile,
+		"-o", "jfr",
+	)
+	if p.cfg.CPU {
+		argv = append(argv, "-e", "itimer")
+	}
+	if p.cfg.Alloc != "" {
+		argv = append(argv, "--alloc", p.cfg.Alloc)
+	}
+	if p.cfg.Lock != "" {
+		argv = append(argv, "--lock", p.cfg.Lock)
+	}
+	argv = append(argv,
+		"start",
+		"--timeout", strconv.Itoa(int(p.interval().Seconds())),
+		strconv.Itoa(p.pid),
+	)
+
+	_ = level.Debug(p.logger).Log("cmd", fmt.Sprintf("%s %s", p.dist.LauncherPath(), strings.Join(argv, " ")))
+	stdout, stderr, err := p.profiler.Execute(p.dist, argv)
+	if err != nil {
+		return fmt.Errorf("asprof failed to run: %w %s %s", err, stdout, stderr)
+	}
+	return nil
+}
+
+func (p *profilingLoop) stop() error {
+	argv := []string{
+		"stop",
+		"-o", "jfr",
+		strconv.Itoa(p.pid),
+	}
+	_ = level.Debug(p.logger).Log("msg", "asprof", "cmd", fmt.Sprintf("%s %s", p.dist.LauncherPath(), strings.Join(argv, " ")))
+	stdout, stderr, err := p.profiler.Execute(p.dist, argv)
+	if err != nil {
+		return fmt.Errorf("asprof failed to run: %w %s %s", err, stdout, stderr)
+	}
+	_ = level.Debug(p.logger).Log("msg", "asprof stopped", "stdout", stdout, "stderr", stderr)
+	return nil
+}
+
+func (p *profilingLoop) update(target discovery.Target) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.target = target
+}
+
+// Close stops profiling this profilingLoop
+func (p *profilingLoop) Close() error {
+	p.cancel()
+	p.wg.Wait()
+	return nil
+}
+
+func (p *profilingLoop) onError(err error) bool {
+	alive := p.alive()
+	if alive {
+		_ = level.Error(p.logger).Log("err", err)
+	} else {
+		_ = level.Debug(p.logger).Log("err", err)
+	}
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.error = err
+	return alive
+}
+
+func (p *profilingLoop) interval() time.Duration {
+	return time.Second * 15 // todo
```

Comment:
I think this should come from the config ?

Reply:
YES!

Reply:
fixed